File: Server/src/main/content/global/skill/production/herblore/potions/UnfinishedPotion.java
package content.global.skill.production.herblore.potions;

import content.global.skill.production.herblore.Herbs;
import core.game.node.item.Item;

/**
 * The enum Unfinished potion.
 */
public enum UnfinishedPotion {
    /**
     * The Guam.
     */
    GUAM(Herbs.GUAM.getProduct(), 3, new Item(91)),
    /**
     * The Rogue's purse.
     */
    ROGUE_PURSE(Herbs.ROGUES_PUSE.getProduct(), 3, new Item(4840)),
    /**
     * The Marrentill.
     */
    MARRENTILL(Herbs.MARRENTILL.getProduct(), 5, new Item(93)),
    /**
     * The Tarromin.
     */
    TARROMIN(Herbs.TARROMIN.getProduct(), 12, new Item(95)),
    /**
     * The Harralander.
     */
    HARRALANDER(Herbs.HARRALANDER.getProduct(), 22, new Item(97)),
    /**
     * The Ranarr.
     */
    RANARR(Herbs.RANARR.getProduct(), 30, new Item(99)),
    /**
     * The Toadflax.
     */
    TOADFLAX(Herbs.TOADFLAX.getProduct(), 34, new Item(3002)),
    /**
     * The Spirit weed.
     */
    SPIRIT_WEED(Herbs.SPIRIT_WEED.getProduct(), 40, new Item(12181)),
    /**
     * The Irit.
     */
    IRIT(Herbs.IRIT.getProduct(), 45, new Item(101)),
    /**
     * The Avantoe.
     */
    AVANTOE(Herbs.AVANTOE.getProduct(), 50, new Item(103)),
    /**
     * The Kwuarm.
     */
    KWUARM(Herbs.KWUARM.getProduct(), 55, new Item(105)),
    /**
     * The Snapdragon.
     */
    SNAPDRAGON(Herbs.SNAPDRAGON.getProduct(), 63, new Item(3004)),
    /**
     * The Cadantine.
     */
    CADANTINE(Herbs.CADANTINE.getProduct(), 66, new Item(107)),
    /**
     * The Lantadyme.
     */
    LANTADYME(Herbs.LANTADYME.getProduct(), 69, new Item(2483)),
    /**
     * The Dwarf weed.
     */
    DWARF_WEED(Herbs.DWARF_WEED.getProduct(), 72, new Item(109)),
    /**
     * The Torstol.
     */
    TORSTOL(Herbs.TORSTOL.getProduct(), 75, new Item(111)),
    /**
     * The Strong weapon poison.
     */
    STRONG_WEAPON_POISON(HerblorePulse.COCONUT_MILK, new Item(6016), 73, new Item(5936)),
    /**
     * The Super strong weapon poison.
     */
    SUPER_STRONG_WEAPON_POISON(HerblorePulse.COCONUT_MILK, new Item(2398), 82, new Item(5939)),
    /**
     * The Strong antipoison.
     */
    STRONG_ANTIPOISON(HerblorePulse.COCONUT_MILK, Herbs.TOADFLAX.getProduct(), 68, new Item(5942)),
    /**
     * The Super strong antipoison.
     */
    SUPER_STRONG_ANTIPOISON(HerblorePulse.COCONUT_MILK, Herbs.IRIT.getProduct(), 79, new Item(5951));

    private final Item base;

    private final Item ingredient;

    private final int level;

    private final Item potion;


    UnfinishedPotion(final Item base, final Item ingredient, final int level, final Item potion) {
        this.base = base;
        this.ingredient = ingredient;
        this.level = level;
        this.potion = potion;
    }


    UnfinishedPotion(final Item ingredient, final int level, final Item potion) {
        this(HerblorePulse.VIAL_OF_WATER, ingredient, level, potion);
    }

    /**
     * Gets ingredient.
     *
     * @return the ingredient
     */
    public Item getIngredient() {
        return ingredient;
    }

    /**
     * Gets level.
     *
     * @return the level
     */
    public int getLevel() {
        return level;
    }

    /**
     * Gets potion.
     *
     * @return the potion
     */
    public Item getPotion() {
        return potion;
    }

    /**
     * Gets base.
     *
     * @return the base
     */
    public Item getBase() {
        return base;
    }

    /**
     * For item unfinished potion.
     *
     * @param item the item
     * @param base the base
     * @return the unfinished potion
     */
    public static UnfinishedPotion forItem(final Item item, final Item base) {
        for (UnfinishedPotion potion : values()) {
            if ((potion.getIngredient().getId() == item.getId() || potion.getIngredient().getId() == base.getId()) && (item.getId() == potion.getBase().getId() || base.getId() == potion.getBase().getId())) {
                return potion;
            }
        }
        return null;
    }

}

File: Server/src/main/content/global/skill/production/herblore/data/potion/UnfinishedPotion.kt
package content.global.skill.production.herblore.data.potion

import content.global.skill.production.herblore.data.Herb
import content.global.skill.production.herblore.item.HerblorePulse
import cfg.consts.Items
import core.game.node.item.Item

/**
 * Unfinished potion
 *
 * @param base The base item used for the potion.
 * @param ingredient The additional ingredient required for the potion.
 * @param level The level of complexity or skill required to create the potion.
 * @param potion The final item that represents the potion.
 * @constructor Unfinished potion
 */
enum class UnfinishedPotion(
    val base: Item, // The base item used for the potion
    val ingredient: Item, // The additional ingredient required for the potion
    val level: Int, // The level of complexity or skill required to create the potion
    val potion: Item // The final item that represents the potion
) {
    /**
     * Guam
     *
     * @constructor Guam
     */
    GUAM(
        ingredient = Herb.GUAM.product,
        level = 3,
        potion = Item(Items.GUAM_POTIONUNF_91)
    ),

    /**
     * Rogue Purse
     *
     * @constructor Rogue Purse
     */
    ROGUE_PURSE(
        ingredient = Herb.ROGUES_PUSE.product,
        level = 3,
        potion = Item(Items.ROGUES_PURSE_POTIONUNF_4840)
    ),

    /**
     * Marrentill
     *
     * @constructor Marrentill
     */
    MARRENTILL(
        ingredient = Herb.MARRENTILL.product,
        level = 5,
        potion = Item(Items.MARRENTILL_POTIONUNF_93)
    ),

    /**
     * Tarromin
     *
     * @constructor Tarromin
     */
    TARROMIN(
        ingredient = Herb.TARROMIN.product,
        level = 12,
        potion = Item(Items.TARROMIN_POTIONUNF_95)
    ),

    /**
     * Harralander
     *
     * @constructor Harralander
     */
    HARRALANDER(
        ingredient = Herb.HARRALANDER.product,
        level = 22,
        potion = Item(Items.HARRALANDER_POTIONUNF_97)
    ),

    /**
     * Ranarr
     *
     * @constructor Ranarr
     */
    RANARR(
        ingredient = Herb.RANARR.product,
        level = 30,
        potion = Item(Items.RANARR_POTIONUNF_99)
    ),

    /**
     * Toadflax
     *
     * @constructor Toadflax
     */
    TOADFLAX(
        ingredient = Herb.TOADFLAX.product,
        level = 34,
        potion = Item(Items.TOADFLAX_POTIONUNF_3002)
    ),

    /**
     * Spirit Weed
     *
     * @constructor Spirit Weed
     */
    SPIRIT_WEED(
        ingredient = Herb.SPIRIT_WEED.product,
        level = 40,
        potion = Item(Items.SPIRIT_WEED_POTIONUNF_12181)
    ),

    /**
     * Irit
     *
     * @constructor Irit
     */
    IRIT(
        ingredient = Herb.IRIT.product,
        level = 45,
        potion = Item(Items.IRIT_POTIONUNF_101)
    ),

    /**
     * Avantoe
     *
     * @constructor Avantoe
     */
    AVANTOE(
        ingredient = Herb.AVANTOE.product,
        level = 50,
        potion = Item(Items.AVANTOE_POTIONUNF_103)
    ),

    /**
     * Kwuarm
     *
     * @constructor Kwuarm
     */
    KWUARM(
        ingredient = Herb.KWUARM.product,
        level = 55,
        potion = Item(Items.KWUARM_POTIONUNF_105)
    ),

    /**
     * Snapdragon
     *
     * @constructor Snapdragon
     */
    SNAPDRAGON(
        ingredient = Herb.SNAPDRAGON.product,
        level = 63,
        potion = Item(Items.SNAPDRAGON_POTIONUNF_3004)
    ),

    /**
     * Cadantine
     *
     * @constructor Cadantine
     */
    CADANTINE(
        ingredient = Herb.CADANTINE.product,
        level = 66,
        potion = Item(Items.CADANTINE_POTIONUNF_107)
    ),

    /**
     * Lantadyme
     *
     * @constructor Lantadyme
     */
    LANTADYME(
        ingredient = Herb.LANTADYME.product,
        level = 69,
        potion = Item(Items.LANTADYME_POTIONUNF_2483)
    ),

    /**
     * Dwarf Weed
     *
     * @constructor Dwarf Weed
     */
    DWARF_WEED(
        ingredient = Herb.DWARF_WEED.product,
        level = 72,
        potion = Item(Items.DWARF_WEED_POTIONUNF_109)
    ),

    /**
     * Torstol
     *
     * @constructor Torstol
     */
    TORSTOL(
        ingredient = Herb.TORSTOL.product,
        level = 75,
        potion = Item(Items.TORSTOL_POTIONUNF_111)
    ),

    /**
     * Strong Weapon Poison
     *
     * @constructor Strong Weapon Poison
     */
    STRONG_WEAPON_POISON(
        base = HerblorePulse.COCONUT_MILK,
        ingredient = Item(Items.CACTUS_SPINE_6016),
        level = 73,
        potion = Item(Items.WEAPON_POISON_PLUSUNF_5936)
    ),

    /**
     * Super Strong Weapon Poison
     *
     * @constructor Super Strong Weapon Poison
     */
    SUPER_STRONG_WEAPON_POISON(
        base = HerblorePulse.COCONUT_MILK,
        ingredient = Item(Items.CAVE_NIGHTSHADE_2398),
        level = 82,
        potion = Item(Items.WEAPON_POISON_PLUS_PLUSUNF_5939)
    ),

    /**
     * Strong Antipoison
     *
     * @constructor Strong Antipoison
     */
    STRONG_ANTIPOISON(
        base = HerblorePulse.COCONUT_MILK,
        ingredient = Herb.TOADFLAX.product,
        level = 68,
        potion = Item(Items.ANTIPOISON_PLUSUNF_5942)
    ),

    /**
     * Super Strong Antipoison
     *
     * @constructor Super Strong Antipoison
     */
    SUPER_STRONG_ANTIPOISON(
        base = HerblorePulse.COCONUT_MILK,
        ingredient = Herb.IRIT.product,
        level = 79,
        potion = Item(Items.ANTIPOISON_PLUS_PLUSUNF_5951)
    );

    constructor(ingredient: Item, level: Int, potion: Item) : this(
        HerblorePulse.VIAL_OF_WATER,
        ingredient,
        level,
        potion
    )

    companion object {
        fun forItem(item: Item, base: Item): UnfinishedPotion? {
            for (potion in values()) {
                if ((potion.ingredient.id == item.id || potion.ingredient.id == base.id) && (item.id == potion.base.id || base.id == potion.base.id)) {
                    return potion
                }
            }
            return null
        }
    }
}
